backends/tc/util.h

```diff
@@ -0,0 +1,141 @@
+
+// String representation of given bit string according to how it is printed by
+// `PrettyPrintBitString`.
+std::string BitStringToString(const std::vector<bool>& bit_string);
+
+// Helper function to convert an unsigned int to bitstring, msb-first
+std::vector<bool> UInt64ToBitString(uint64_t n);
+
+// Helper function to convert an unsigned int to bitstring with given size,
+// msb-first. If n is too large to fit in `size` bits, then the most-significant
+// bits are discarded. If n requires fewer bits than `size` to represent, then
+// it is padded right, and the most-significant bits are set to 0. This behavior
+// is same as the behavior of HexStringToBitString.
+std::vector<bool> UInt64ToBitString(uint64_t n, size_t size);
+
+// Helper function to convert a hexadecimal string to bitstring, msb-first. if
+// the hex string is too long, then the most-significant bits are discarded. If
+// it is too short, then it is padded right (used for the least-significant
+// bits) and most-significant bits are set to zero.
+std::vector<bool> HexStringToBitString(size_t size,
+                                       absl::string_view hex_string);
+
+// Convert a sized string to bit string. This function expects a string of the
+// form SIZEwVALUE where VALUE is a hexadecimal string
+// prefixed by 0x. Support for decimal representation could be added in the
+// future.
+//
+// Its padding behavior is same as that of `HexStringToBitString`.
+absl::StatusOr<std::vector<bool>> SizedStringToBitString(
+    absl::string_view string_with_size);
+
+// Parse given string as a number. This method is a more flexible wrapper around
+// the standard library methods for parsing integers, and it returns
+// `absl::StatusOr` instead of an error code or throwing an exception.
+absl::StatusOr<uint32_t> ParseUInt32(absl::string_view string);
+
+// A function object for `operator<<` where left-hand side is a stream. The
+// standard library does not provide a named functional object for `operator<<`.
+template <class Lhs, class Rhs>
+struct WriteToStream {
+  auto operator()(Lhs&& lhs, Rhs&& rhs) const
+      -> decltype(std::forward<Lhs>(lhs) << std::forward<Rhs>(rhs)) {
+    return std::forward<Lhs>(lhs) << std::forward<Rhs>(rhs);
+  }
+};
+
+template <class Lhs>
+struct WriteToStream<Lhs, void> {};
+
+template <class Rhs>
+struct WriteToStream<void, Rhs> {};
+
+template <>
+struct WriteToStream<void, void> {};
+
+// A higher-order function that implements fold (reduce) left on the type
+// level. This is a substitute for fold expressions in C++17.
+//
+// The binary operator BinaryOp is assumed to have the type signature
+// `template<typename First, typename Second> First(First, Second)` modulo
+// lvalue/rvalue forwarding. That is, it should produce a functional object
+// class implements a binary operator of type `(First * Second) -> First`
+template <template <class, class> class BinaryOp, class Arg>
+Arg FoldLeft(Arg&& arg) {
```

Comment:
where is this used?

Reply:
It is used by `ReportErrorAt` in `ir_builder.h` [defined here](https://github.com/p4lang/p4c/pull/2898/files/535a49c6e0811517c5c0dbdd15f763581050cbeb#diff-c150abd4131b7e93d9dbca2ad1137b5a6806edc312ed81ee75d05ab2b049f86dR149).

`ReportErrorAt` uses it to print the error message in a type-safe manner using `operator<<` instead of format strings.

Reply:
I saw you also left a comment on `ReportErrorAt`, I also responded there to clarify which part of it this function is used for.